<a href="https://colab.research.google.com/github/RichardOmbina/MTN-Strategy/blob/master/Richard_Ombina_MTN_strategy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Research Questions**
1 Which ones were the most used city for the three days?
2 Which cities were the most used during business and home hours?
3 Most used city for the three days?
  
The telecom data provided for this project is only a sample ( i.e. for only three days).   

cells_geo_description.xlsx [Link] (Links to an external site.)
cells_geo.csv [Link] (Links to an external site.)
CDR_description.xlsx [Link] (Links to an external site.)
CDR 20120507 [http://bit.ly/TelecomDataset1] (Links to an external site.)
CDR 20120508 [http://bit.ly/TelecomDataset2] (Links to an external site.)
CDR 20120509 [http://bit.ly/TelecomDataset3]

In [9]:
# Importing libraries to be used
import pandas as pd
import numpy as np


In [11]:
#Loading datasets
cells_df = pd.read_csv('cells_geo.csv', sep=";", encoding='utf-8')

column_names = ['product', 'bill',	'datetime','cell_on_site','caller_anonymized', 'recipient_anonymized', 'country_a','country_b','cell_id', 'site_id']
#T1,T2,T3 are telcome datasets
T1 = pd.read_csv('Telcom_dataset.csv', encoding='utf-8', names=column_names, header=0)
T2 = pd.read_csv('Telcom_dataset2.csv', encoding='utf-8',names = column_names, header=0)
T3 = pd.read_csv('Telcom_dataset3.csv', encoding='utf-8', names = column_names, header=0)

In [15]:
#confirming the dataset is loaded
cells_df

Unnamed: 0    VILLES      STATUS  ...          AREA     CELL_ID   SITE_CODE
0              0    ADJAME  In Service  ...           NaN  fbb2e75af3  0bb492ef17
1              1    ADJAME  In Service  ...           NaN  1502501d78  0bb492ef17
2              2    ADJAME  In Service  ...           NaN  0c2e182a0c  0bb492ef17
3              3   ASSINIE  In Service  ...           NaN  69bdfc69fd  8a1e6ce31f
4              4   ASSINIE  In Service  ...           NaN  7a1029d2ab  8a1e6ce31f
...          ...       ...         ...  ...           ...         ...         ...
3969        3969  ZUENOULA  In Service  ...  CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970        3970  ZUENOULA  In Service  ...  CENTRE NORTH  3ae1708153  f91ed8d2e9
3971        3971  ZUENOULA  In Service  ...  CENTRE NORTH  cb71f8a29c  61bee9c70c
3972        3972  ZUENOULA  In Service  ...  CENTRE NORTH  d4a6a0b645  61bee9c70c
3973        3973  ZUENOULA  In Service  ...  CENTRE NORTH  f9f3e42589  61bee9c70c

[3974 rows x 12 columns]

In [18]:
# Droping unnamed column
cells_df.drop('Unnamed: 0', axis=1, inplace=True)

KeyError: ignored

In [20]:
#confirming column is droped
cells_df.head(10)

VILLES      STATUS LOCALISATION  ... AREA     CELL_ID   SITE_CODE
0    ADJAME  In Service      ABIDJAN  ...  NaN  fbb2e75af3  0bb492ef17
1    ADJAME  In Service      ABIDJAN  ...  NaN  1502501d78  0bb492ef17
2    ADJAME  In Service      ABIDJAN  ...  NaN  0c2e182a0c  0bb492ef17
3   ASSINIE  In Service    INTERIEUR  ...  NaN  69bdfc69fd  8a1e6ce31f
4   ASSINIE  In Service    INTERIEUR  ...  NaN  7a1029d2ab  8a1e6ce31f
5   ASSINIE  In Service    INTERIEUR  ...  NaN  0b4f56438f  8a1e6ce31f
6  KRIKOREA  In Service    INTERIEUR  ...  NaN  041d2504f5  260dbfedc5
7     ABOBO  In Service      ABIDJAN  ...  NaN  a70c1d263b  56306cc7ca
8     ABOBO  In Service      ABIDJAN  ...  NaN  e2d04ea883  56306cc7ca
9     ABOBO         NaN      ABIDJAN  ...  NaN  ee224cd27e  56306cc7ca

[10 rows x 11 columns]

In [21]:
# droping irrelevant columns
droped = ['cell_on_site','country_a','country_b']
T1.drop(droped, axis=1,inplace=True)
T2.drop(droped, axis=1,inplace=True)
T3.drop(droped, axis=1,inplace=True)

In [22]:
T1.head(10)

product  bill  ...     cell_id     site_id
0   Voice     0  ...  /TJNe+Mmtu  /TJNe+Mmtu
1     sms     0  ...  /+cKZKIp41  /+cKZKIp41
2     sms    25  ...  /u0FSD+ahi  /u0FSD+ahi
3   Voice    56  ...  +cNeJzsTp3  +cNeJzsTp3
4   Voice    14  ...  051KhYDCpv  051KhYDCpv
5    data     0  ...  /lrUXVGVcN  /lrUXVGVcN
6   Voice     0  ...  +38u2u/rfx  +38u2u/rfx
7   Voice     0  ...  09UjmmY1Ds  09UjmmY1Ds
8   Voice     0  ...  0DeimCpZOc  0DeimCpZOc
9     sms     0  ...  01MDiY5F89  01MDiY5F89

[10 rows x 7 columns]

In [23]:
# Check for missing values
cells_df.isnull().sum()

VILLES           0
STATUS          67
LOCALISATION     0
DECOUPZONE       0
ZONENAME         6
LONGITUDE        0
LATITUDE         0
REGION           0
AREA            23
CELL_ID          0
SITE_CODE        0
dtype: int64

In [24]:
# Droping duplicates
cells_df.drop_duplicates(inplace=True)
T1.drop_duplicates(inplace=True)
T2.drop_duplicates(inplace=True)
T3.drop_duplicates(inplace=True)

In [25]:
# Joining dataframe T1,T2,T3,
joined_df = pd.concat([T1,T2,T3], axis=0, join='outer',ignore_index=True)
joined_df.head(10)

product  bill  ...     cell_id     site_id
0   Voice     0  ...  /TJNe+Mmtu  /TJNe+Mmtu
1     sms     0  ...  /+cKZKIp41  /+cKZKIp41
2     sms    25  ...  /u0FSD+ahi  /u0FSD+ahi
3   Voice    56  ...  +cNeJzsTp3  +cNeJzsTp3
4   Voice    14  ...  051KhYDCpv  051KhYDCpv
5    data     0  ...  /lrUXVGVcN  /lrUXVGVcN
6   Voice     0  ...  +38u2u/rfx  +38u2u/rfx
7   Voice     0  ...  09UjmmY1Ds  09UjmmY1Ds
8   Voice     0  ...  0DeimCpZOc  0DeimCpZOc
9     sms     0  ...  01MDiY5F89  01MDiY5F89

[10 rows x 7 columns]

In [29]:
merged_df = pd.merge(joined_df, cells_df, how='inner', left_on=['site_id'], right_on=['SITE_CODE'])
merged_df.head()

product  bill               datetime  ...          AREA     CELL_ID   SITE_CODE
0     sms     0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  24ccdda3d5  1b5540c02d
1     sms     0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  ebfc58af70  1b5540c02d
2     sms     0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  ba8ce6e4ef  1b5540c02d
3     sms     0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  c2545171dc  1b5540c02d
4     sms     0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  f15bd5483f  1b5540c02d

[5 rows x 18 columns]

In [31]:
merged_df.shape

(49702, 18)

In [74]:
#drop columns 
to_drop = ['recipient_anonymized','cell_id','CELL_ID','SITE_CODE']
merged_df.drop(to_drop, axis=1, inplace=True)


KeyError: ignored

In [35]:
merged_df

product  bill  ...             REGION           AREA
0         sms     0  ...  VALLEE DU BANDAMA   CENTRE NORTH
1         sms     0  ...  VALLEE DU BANDAMA   CENTRE NORTH
2         sms     0  ...  VALLEE DU BANDAMA   CENTRE NORTH
3         sms     0  ...  VALLEE DU BANDAMA   CENTRE NORTH
4         sms     0  ...  VALLEE DU BANDAMA   CENTRE NORTH
...       ...   ...  ...                ...            ...
49697   Voice    17  ...             LAGUNE  ABIDJAN NORTH
49698   Voice    17  ...             LAGUNE  ABIDJAN NORTH
49699     sms     0  ...  VALLEE DU BANDAMA   CENTRE NORTH
49700     sms     0  ...  VALLEE DU BANDAMA   CENTRE NORTH
49701     sms     0  ...  VALLEE DU BANDAMA   CENTRE NORTH

[49702 rows x 14 columns]

In [36]:
#checking change in dataframe size
merged_df.shape

(49702, 14)

In [37]:
# Which regions have the highest frequency of calls?
merged_df['REGION'].value_counts()

LAGUNE               32547
LACS                  2408
BAS SASSANDRA         2327
HAUT SASSANDRA        2277
VALLEE DU BANDAMA     1551
SUD BANDAMA           1518
SAVANE                 980
FROMAGER               904
WORODOUGOU             792
18 MONTAGNES           765
MARAHOUE               618
AGNEBY                 581
NZI COMOE              564
MOYEN CAVALLY          528
DENGUELE               395
ZANZAN                 330
MOYEN COMOE            315
SUD COMOE              203
BAFING                  99
Name: REGION, dtype: int64

**The top three regions with highest calls are**

LAGUNE               32547
LACS                  2408
BAS SASSANDRA         2327




In [38]:
merged_df['ZONENAME'].value_counts()

ZONE C - Abidjan_OUEST     10296
ZONE E - Abidjan_NORD       6660
ZONE B - Grand-SUD          5545
ZONE B - Abidjan_EST        4838
ZONE D - Abidjan_SUD        4642
ZONE D - Grand-OUEST        4314
ZONE A - Abidjan_CENTRE     4132
ZONE A - Grand-NORD         3928
ZONE E - SUD-OUEST          2993
ZONE C - Grand-EST          1748
ZONE D - Abidjan_EST         186
OKAHOTEL                     120
NPONON                        72
WILLIAMSVILLE2                40
ZONE C - Abidjan_EST          27
ZONE A - Abidjan_OUEST        21
PALMABINGERVILLE              21
ZONE B - Grand-EST            18
ZONE B - Abidjan_OUEST        18
ZONE A - Grand-SUD            18
ZONE C - Abidjan_SUD          15
ZONE A - Grand-EST             6
ZONE B - Abidjan_SUD           6
ZONE D - Grand-SUD             6
AGBANNEW                       3
ZONE B - Abidjan_Nord          3
ZONE B - Abidjan-EST           3
ZONE B - Grand-NORD            3
ZONE B - Abidjan_CENTRE        2
Name: ZONENAME, dtype: int64


**The top fivezones with highest calls are**

**ZONE C** - Abidjan_OUEST     10296

**ZONE E** - Abidjan_NORD       6660

**ZONE B** - Grand-SUD          5545

**ZONE B** - Abidjan_EST        4838

**ZONE D** - Abidjan_SUD        4642

In [45]:
#defining high traffic regions
high_traffic_regions = ['LAGUNE','LACS','BAS SASSANDRA','HAUT SASSANDRA','VALLEE DU BANDAMA','SUD BANDAMA']

region_df = merged_df[merged_df['REGION'].isin(high_traffic_regions)]
cities_regions = region_df['VILLES'].value_counts()
high_cities_regions = cities_regions[cities_regions > 500]
high_impact_cities = high_cities_regions | high_cities_regions
cities_df = merged_df[merged_df['VILLES'].isin(high_impact_cities.index)]
services = cities_df['product'].value_counts(normalize=True)
services


sms      0.61448
Voice    0.38552
Name: product, dtype: float64

**SMS is the most used service**

In [49]:
#product bringing more money
profits = cities_df.groupby(['product'])['bill'].sum()
profits

product
Voice    815315
sms      188372
Name: bill, dtype: int64

**Voice fetches more money**

In [52]:
cities_df.groupby(['VILLES'])['bill'].sum().sort_values(ascending=False)

VILLES
YOPOUGON        202167
COCODY          177745
ABOBO           124210
KOUMASSI         71356
TREICHVILLE      65557
MARCORY          59913
ADJAME           59455
YAMOUSSOUKRO     45090
SAN-PEDRO        44137
PORT-BOUET       35786
DALOA            34704
BOUAKE           33039
DIVO             23676
ATTECOUBE        19116
SOUBRE            7736
Name: bill, dtype: int64

**Which cities are bringing more money**

**VILLES**

**YOPOUGON**       202167

**COCODY**          177745

**ABOBO**        124210

In [54]:
# Which city brings in more money per product?
product_bills = cities_df.groupby(['VILLES','product'])['bill'].sum().reset_index()
product_bills

VILLES product    bill
0          ABOBO   Voice   95781
1          ABOBO     sms   28429
2         ADJAME   Voice   45925
3         ADJAME     sms   13530
4      ATTECOUBE   Voice   14266
5      ATTECOUBE     sms    4850
6         BOUAKE   Voice   28302
7         BOUAKE     sms    4737
8         COCODY   Voice  157920
9         COCODY     sms   19825
10         DALOA   Voice   26044
11         DALOA     sms    8660
12          DIVO   Voice   16701
13          DIVO     sms    6975
14      KOUMASSI   Voice   55781
15      KOUMASSI     sms   15575
16       MARCORY   Voice   49085
17       MARCORY     sms   10828
18    PORT-BOUET   Voice   27907
19    PORT-BOUET     sms    7879
20     SAN-PEDRO   Voice   39606
21     SAN-PEDRO     sms    4531
22        SOUBRE   Voice    6561
23        SOUBRE     sms    1175
24   TREICHVILLE   Voice   58932
25   TREICHVILLE     sms    6625
26  YAMOUSSOUKRO   Voice   37590
27  YAMOUSSOUKRO     sms    7500
28      YOPOUGON   Voice  154914
29      YOPOUGON     sms   47253

In [62]:
product_bills['bill'].max()

157920

In [64]:
product_bills.loc[product_bills['bill'] == 157920]

VILLES product    bill
8  COCODY   Voice  157920

**COCODY** brings in more money per product.